### DISCUSSION TITLE
Can I use this library for sharing entities between users?

### DISCUSSION THREAD

**Discussion by nilskretschmer** [spatie/laravel-permission#1813] [Q&A]
Hi,

I really like this library and already read through the docs. Currently I am trying to find out if I can use it for some kind of sharing mechanism. I am not sure if there will be a performance problem because I would need a lot of custom permissions.

Think of it like a simple todolist app:
User A wants to share todolist (id: 1) with user B. So both users get permissions to access todolist (id: 1). Maybe there will be more specific permissions what users can do with shared todolists (edit, delete, share..).

Problem:
N todolists require n permissions (+ specific permissions as mentioned above).

Can I use this library for that kind of mechanism or will there be performance problems? Should I implement it differently (like defining permission tables)?

Thanks in advance!

**Comment by erikn69** [comment]
https://spatie.be/docs/laravel-permission/v4/best-practices/roles-vs-permissions

Maybe you can get my idea for this (#1804)

shared todolist (id: 1) must be a pivot on `model_has_permissions`(maybe on `model_has_roles` too)

So now you can assign the same permission for granular shared todolist, 

On #1782 i better explain my method

>performance problems

I don´t know, i believe that you don't have many permissions, so no cache problems, but if you have a lot todolist maybe if you make relation load all at start

**Comment by nilskretschmer** [comment]
I am not sure if I get your idea right: Do you mean I still need additional pivot tables for the shared todolists? aka user_todolist_table which holds user_id and todolist_id?

**Reply by nilskretschmer** [reply]
In addition to your answer: Where is the info stored which todolist is associated with which user in your example? So let's assume my API route /api/todolists returns all todolists the user is owner or at least read-permitted. In the controller I'd need to retrieve only those specific todolists from the DB. Please answer back if something's not clear enough. My english isn't the best :)

**Reply by erikn69** [reply]
The package is not made to "share entities between users", I'm just trying to give you ideas to achieve it, it needs work on your side, because I have not done something like that. (There could be a package that already does this)

>I still need additional pivot tables for the shared todolists? aka user_todolist_table which holds user_id and todolist_id?

Maybe not, because you already has `model_has_permissions` with that pivot, but it depends on your solution

>Where is the info stored which todolist is associated with which user in your example?

 `model_has_permissions`

>returns all todolists the user is owner or at least read-permitted. I'd need to retrieve only those specific todolists from the DB

If you have a new pivot, It does exactly that

You can make a middleware to set wich todolists id user is trying to access, the relationship loads based on the id that is in the route(for performance)
